internal/internal_decision_state_machine.go

```diff
@@ -862,6 +862,25 @@ func (h *commandsHelper) addCommand(command commandStateMachine) {
 	h.incrementNextCommandEventID()
 }
 
+// This really should not exist, but is unavoidable without totally redesigning the Go SDK to avoid
+// doing event number counting. EX: Because a workflow execution cancel requested event calls a callback
+// on timers that immediately cancels them, we will queue up a cancel timer command even though that timer firing
+// might be in the same workflow task. In practice this only seems to happen during unhandled command events.
+func (h *commandsHelper) removeCommand(commandID commandID) {
```

Comment:
This should be generalized to handle other cases not just for timer:
strawman proposal:
```
func (h *commandsHelper) removeCommand(id commandID) bool {
	if elem, ok := h.commands[id]; ok {
		h.orderedCommands.Remove(elem)
		delete(h.commands, id)
		cmd := elem.Value.(commandStateMachine)
		// cmd is gone, now we need to make sure we don't try to cancel that cmd
		switch cmd := cmd.(type) {
		case *timerCommandStateMachine:
			cancelCmd := makeCommandID(commandTypeCancelTimer, cmd.getID().id)
			if h.removeCommand(cancelCmd) {
				h.commandsCancelledDuringWFCancellation--
			}
		case *activityCommandStateMachine:
			cancelCmd := makeCommandID(commandTypeRequestCancelActivityTask, cmd.getID().id)
			if h.removeCommand(cancelCmd) {
				h.commandsCancelledDuringWFCancellation--
			}
		// add other cases that make sense here.
		}
		return true
	}
	return false
}
```

Then, any where in this file that calls orderedCommands.Remove() should be replaced to call this removeCommand() method.

Reply:
I'm inclined to contain the blast of this to just the specific case that needs to be fixed since it's such a lame patch. Although this can technically happen to activities I'm worried making more use of it is just going to lead to more potential problems down the line when we really want to just re-do all of this. I can make it apply to both if we really want though

Reply:
I still think this should be handled here in general way, not in handleTimerClosed(). Agree that if you want to limit the case to only handle timer for now.

Reply:
Actually I think we should fix all cases here. Because each case would cause non-deterministic and block user and it is a time bomb waiting to exploded.

Reply:
K, sounds good. Thanks for the sample code!

Reply:
So, turned out this didn't work since we only have the information that this should be done inside the closed event handlers